### PR TITLE
Recreate build pipeline within tectonic console

### PR DESCRIPTION
**Pull request by benjaminapetersen**
- Per task https://jira.coreos.com/browse/CONSOLE-453

w/@sg00dwin, bah, wasn't going to squash your commit into mine, sorry about that. 😦 

Looking over a few more details from the old pipeline's flow, then I'll remove the WIP.  Any feedback welcome.

### CODE REVIEW

**Comment by benjaminapetersen** [comment]
Pending:
screen shot 2018-05-29 at 10 49 54 am

Build in progess:
screen shot 2018-05-29 at 10 51 24 am

A complete flow:
screen shot 2018-05-29 at 10 52 55 am

File: frontend/public/components/build-pipeline.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    <BuildAnimationCircle />
+  </div>
+}
+
+const JenkinsInputUrl: React.SFC<JenkinsInputUrlProps> = ({resource, stage}) => {
+  let pending = stage && (stage.status === 'PAUSED_PENDING_INPUT');
+  let buildUrl = getJenkinsBuildURL(resource);
+  if(pending) {
+    return <div className="build-pipeline__stage-actions text-muted">
+      <a href={buildUrl}>Input Required</a>
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]:
```
target="_blank" rel="noopener noreferrer"
```

File: frontend/public/components/build-pipeline.tsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    </div> 
+  }
+  return null;
+}
+
+const BuildStageTimestamp = ({timestamp}) => <div className="build-pipeline__stage-time text-muted">
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: types?

File: frontend/public/components/build-pipeline.tsx [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    <div className="build-pipeline__container">
+      <div className="build-pipeline__stages">
+        { pipelinesStarted ? null : <StagesNotStarted /> }
+        { stages.map((stage => <BuildStage resource={resource} stage={stage} key={stage.id} /> )) }
+      </div>
+    </div>
+  </div>;
+};
+
+// NOTE: is not instantiated, so need this wrapper?
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: I'd leave off the comment since it applies to all tsx files

File: frontend/public/components/build.tsx [hunk 4]
```diff
@@ -38,6 +39,11 @@ export const BuildsDetails: React.SFC<BuildsDetailsProps> = ({obj: build}) => {
   const duration = formatBuildDuration(build);
 
   return <div className="co-m-pane__body">
+    <div className="row">
```
Comment (spadgett) [inline frontend/public/components/build.tsx]: Maybe it's OK, but this is going to add some empty DOM nodes if it's not actually a build pipeline

File: frontend/public/components/build-pipeline.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+const getJenkinsStatus = (resource: any) => {
+  const status = _.get(resource, ['metadata', 'annotations', 'openshift.io/jenkins-status-json']);
+  return status && JSON.parse(status);
+};
+const getJenkinsLogURL = (resource: any) => _.get(resource, ['metadata', 'annotations', 'openshift.io/jenkins-log-url']);
+const getJenkinsBuildURL = (resource: any) => _.get(resource, ['metadata', 'annotations', 'openshift.io/jenkins-build-uri']); 
+
+
+const BuildSummaryStatusIcon: React.SFC<BuildSummaryStatusIconProps> = ({status}) => {
+  let statusClass = _.lowerCase(status);
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: also polishing up for next push.

File: frontend/public/components/build-pipeline.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
+      <span aria-hidden="true">
+        <span className={`fa ${icon} fa-fw`}></span>
+      </span>
+    </span>
+
+    <span className={`build-pipeline__status-icon ${hide}`}>
+      <span className="fa fa-refresh fa-spin" aria-hidden="true"></span>
+    </span>
+  </React.Fragment>
```
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/build-pipeline.tsx]: It does: https://s3.amazonaws.com/teamui-jenkins/jenkins-console-build-6290/Tests.log

Try `yarn run lint`

File: frontend/public/components/build-pipeline.tsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+</div>;
+
+
+const BuildSummaryTimestamp: React.SFC<BuildSummaryTimestampProps> = ({timestamp}) => <span className="build-pipeline__timestamp text-muted">
+  { fromNow(timestamp) }
+</span>
+
+const BuildPipelineSummary: React.SFC<BuildPipelineSummaryProps> = ({resource}) => {
+  const { name, namespace } = resource.metadata;
+  const buildNumber: number = getBuildNumber(resource);
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: will do, that makes more sense.

File: frontend/public/components/build-pipeline.scss [hunk 8]
```diff
@@ -0,0 +1,433 @@
+$pipeline-aborted-color: #d1d1d1;
+$pipeline-border-color: #d1d1d1;
+$pipeline-failed-color: #c00;
+$pipeline-font-base: 12px;
+$pipeline-in-progress-color: #0088ce;
+$pipeline-new-color: #bee1f4;
+$pipeline-pending-color: #d1d1d1;
+$pipeline-success-color: #3f9c35;
```
[the remark this answers is not in the document]
Comment (sg00dwin) [reply frontend/public/components/build-pipeline.scss]: We can. I gave @benjaminapetersen the hex incorrectly.

@benjaminapetersen can you replace with the following vars.
```
$pipeline-aborted-color: $color-pf-black-300;
$pipeline-border-color: $color-pf-black-300;
$pipeline-failed-color: $color-pf-red-100;
$pipeline-in-progress-color: $color-pf-blue-400;
$pipeline-new-color: $color-pf-blue-100;
$pipeline-pending-color: $color-pf-black-300;
$pipeline-success-color: $color-pf-green-400;
$pipeline-require-attention-color: $color-pf-gold-400;
```

File: frontend/public/components/build-pipeline.scss [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+  flex: 1 1 auto;
+  overflow: hidden;
+}
+.build-pipeline__stages {
+  display: flex;
+  flex-wrap: wrap;
+  height: 100%;
+  padding: 0 ($pipeline-padding / 2);
+}
+.build-pipeline__stage {
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: @rhamilto is that standard in this repo?  or have they followed another convention?
(asking cuz I'm still not a huge fan of alpha, ha.)

File: frontend/public/components/build-pipeline.scss [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+    line-height: 1;
+    position: absolute;
+    right: 0;
+    text-align: center;
+  }
+  // hide arrow on last stage
+  &:last-child:before {
+    display: none;
+  }
+  &.build-pipeline__stage--none {
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: @sg00dwin I'll defer to you on the style

File: frontend/public/components/build-pipeline.scss [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+}
+.build-pipeline__stage-time, .build-pipeline__stage-actions {
+  margin-top: 12px;
+  &--in-progress {
+    color: #777;
+  }
+}
+
+.build-pipeline__status-icon {
+  &--complete {
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: @sg00dwin

File: frontend/public/components/build-pipeline.tsx [hunk 12]
```diff
@@ -0,0 +1,173 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import { Link } from 'react-router-dom';
+import { resourcePath } from './utils';
+import { fromNow } from './utils/datetime';
+
+const getBuildNumber = (resource: any) => _.get(resource, ['metadata', 'annotations', 'openshift.io/build.number']);
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Perfect, I'd rather use `K8sResourceKind`.

File: frontend/public/components/build-pipeline.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (rhamilto) [reply frontend/public/components/build-pipeline.scss]: It is not (I'm not sure there is one), but it does appear to be in use within this partial.

**Comment by benjaminapetersen** [comment]
@kans yeah let me see if I can hit all of those:

1.  I'd love to see this for every PR, actually.   A screenshot & "how do I use this?" is great for anyone reviewing, as well as historical documentation. 

The [Openshift QE V3 Test Files](https://github.com/openshift-qe/v3-testfiles) repo has quite a few bits of `yaml` and `json` to create resources in a variety of states.  There are a handful of `Pipelines` options, for the sample below, I'm pulling the first one that came up:

```bash
# create a project
oc process -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/99d26e235b093431bff70e9edffc524559a9d664/pipeline/ui-pipeline-stage.yaml | oc create -f -
```
Then you should be able to navigate to a `build` page, see the `pipeline`, click `action` -> `rebuild` to run again, etc.

2.  Integration tests for animations... I'd probably expect an exceptionally good set of unit tests & integration tests before worrying about testing a css animation.  Not sure there is a ton value in pursuing.  

3.  Seems like an edge case, I'd prob trust a manual test for a css animation & worry about automated testing for all kinds of other things before worrying about a test case for this.  Could be an enormous rabbit hole.  Open to feedback, of course.  

4.  Definitely open to re-evaluation.  This is not the full pipeline experience from the old console. I'd say comment away on the items in the PR that are worth discussion and lets talk (its essentially 1 CSS file and 1 JS file).  I think both @sg00dwin and I would be happy to kick around chagnes, optimizations, etc.

Thanks for the feedback!

**Comment by spadgett** [comment]
> In the not so distant past, we have seen trivially simple CSS animations use an entire CPU core in Chrome (a single blinking indicator). What does this do to performance?

@kans How have you profiled animations in the past?

I feel like, specifically for pipelines, they add a lot to the user experience. It's immediately apparent what stage is active and when the pipeline moves to the next stage.

File: frontend/public/components/build-pipeline.tsx [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+const getStages = (status: any) => status && status.stages || [];
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Yup but I can update.

File: frontend/public/components/build-pipeline.tsx [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    new: 'fa-hourglass-o',
+    pending: 'fa-hourglass-half',
+    running: 'fa-refresh fa-spin',
+    complete: 'fa-check-circle',
+    failed: 'fa-times-circle'
+  })[statusClass];
+  let hide = icon ? 'hide' : '';
+  console.log('status is: ', status, statusClass);
+  return <React.Fragment>
+    <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Would this be to conditionally add the `build-pipeline__status-icon--${status}` class, ONLY if there is a valid status?  If I could do this:

```javascript
  const klass = classNames(
    'build-pipeline__status-icon',
    {`build-pipeline__status-icon--${statusClass}` : !!statusClass}
  );
```
I think it would make sense, but the ``backtick`` operator isn't valid for creating an object key, so it doesn't help me take advantage of the conditional logic provided by `className`.

File: frontend/public/components/build-pipeline.tsx [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    <BuildSummaryTimestamp timestamp={resource.metadata.creationTimestamp} />
+    <BuildLogLink resource={resource} />
+  </div>
+};
+
+const BuildAnimationLine: React.SFC = () => {
+  return <div className="build-pipeline__animation-line"></div>
+};
+
+const BuildAnimationCircle: React.SFC = () => {
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Will do.

**Comment by spadgett** [comment]
cc @openshift/team-ux-review

**Comment by beanh66** [comment]
Thanks @spadgett, we will discuss in our design meeting today and will follow up this afternoon with comments.

**Comment by serenamarie125** [comment]
@spadgett sorry bout the post 24 hour reply ... just connected with @sspeiche and others to get more background so we could provide a more thoughtful response.

The current design team was not involved with the pipeline design, although in our opinion we agree that it provides an positive user experience.

To provide a bit of history, we needed to have a CI/CD story for our platform...Jenkins is the #1 workflow/pipeline open source tool.  The goal was to provide an integrated experience.  The design "mirrored" the state in Build objects in OpenShift from Jenkins instead of forcing people to go back and forth.  The original design process was a very long process driven by designers who did engage with customers during the design process.  The design process was iterative.

The general feedback from customers has been that they really like it.  There are some specific requests to improve on it, but these have yet to be addressed.  Note that this is still a relatively new feature in the rate our customers roll out updates. So usage is picking up more now ... 

One question going forward would be more around role based access.  This feature would definitely be expected by an "app/dev" role, but unsure for admins.

@robszumski do you have additional thoughts?

**Comment by serenamarie125** [comment]
One more note, OpenShift.io uses the same exact pipeline representation as OpenShift

**Comment by tlwu2013** [comment]
@beanh66, @cshinn, and I briefly chatted about the design yesterday in our sync. 
We feel it seems fine to keep the current design to matchup the experience on App Console side.
So I think from design standpoint, this seems to be good go. But I can't comment on engineering/testing point of view, so I'll defer that to dev team for extended discussions.
And if this PR did introduce serious performance throttle, I am open to exploring tweaks to the design to see if we can improve that.

BTW, just FYI, here's the screen capture I did from the youtube provided by @beanh66 for showing the interaction OpenShift Console side:
https://drive.google.com/file/d/1NzXtv5sLSvbUruF4SNru-NmR8eYwThW7/view?usp=sharing

**Comment by spadgett** [comment]
@benjaminapetersen @sg00dwin do you feel like this is ready for review?

@rhamilto can you review the styles?

**Comment by benjaminapetersen** [comment]
Running it again locally today, but the `scss` doesn't seem to build correctly (i think its something local, not that there is a `scss` error).

**Comment by benjaminapetersen** [comment]
Ok, local cache. solved.

Yeah, @spadgett @rhamilto I think its safe to give it a second round of review.  @sg00dwin agreed?

File: frontend/public/components/build-pipeline.scss [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    &:before {
+      bottom: auto;
+      content: '\2192';
+      left: auto;
+      right: 10px;
+      top: 37%;
+    }
+  }
+}
+
```
Comment (benjaminapetersen) [inline frontend/public/components/build-pipeline.scss]: Untested, but since we have that open topic regarding `sass` loops, An alternative could be to use a `sass` map, something like this:

```scss
// can also use a map, with key-value pairs
$pipeline-stages-per-row: (
  small: (
    column-no: 4,
    media-query-px: $screen-md-min
  )
  medium: (
    column-no: 4,
    media-query-px: $screen-lg-min
  )
  large: (
    column-no: 4,
    media-query-px: $screen-lg-min + 200
  )
  x-large: (
    column-no: 4,
    media-query-px: $screen-lg-min + 400
  )
)

// and use the map-get() function to get value by key...
@each $size, $map in $pipeline-stages-per-row {
  @media (min-width: map-get($map, media-query-px)) {
    .build-pipeline__stage {
      width: (100% / map-get($map, column-no));
    }
  }
}
```
`map-get($map, the-key-to-get)` is clearer than `nth($screenSize, someIndex)`, but the code is a bit longer.

File: frontend/public/components/build-pipeline.scss [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+  to {opacity: 1}
+}
+
+@keyframes pipeline-stage-fadeOut {
+  to {background-color: transparent}
+}
+
+.build-pipeline {
+  border: 1px solid $pipeline-border-color;
+	font-size: $pipeline-font-base;
```
Comment (rhamilto) [inline frontend/public/components/build-pipeline.scss]: Nit: spacing

**Comment by benjaminapetersen** [comment]
@rhamilto @sg00dwin  I'm good with choosing an organizational pattern going forward.  Still not fond of how alphabetical breaks up logical groupings like `height` and `width`, or `top`,`bottom`,`left`,`right`, etc, but any organization is better than random.

**Comment by benjaminapetersen** [comment]
@rhamilto opinion on the loops?  Post refactor, prob worth assessing if its really an improvement or no.

File: frontend/public/components/build-pipeline.scss [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+  top: $pipeline-circle-border-width;
+  width: ($pipeline-circle-diameter - $pipeline-circle-border-width * 2);
+}
+.build-pipeline__circle-inner-fill {
+  border-radius: 50%;
+  box-sizing: border-box;
+  height: 100%;
+  opacity: 0;
+  width: 100%;
+}
```
Comment (benjaminapetersen) [inline frontend/public/components/build-pipeline.scss]: Just an observation, since we've been kicking around the topic.  Most of the other existing SCSS files seem to hover around 50 lines or less.  There are a couple in the range of 100.  

350 lines is pretty big.

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: Per the file size, I'm wondering if it would be good to break the elements down.  I'm partial to the pattern I see in a lot of React apps where you have something like:

```bash 
/public
  /components
    /build-pipeline
      |_ build-pipeline.js 
      |_ build-stage.js 
      |_ etc.
```
And do an 

```JavaScript
// build-pipeline.js
import './build-pipeline.scss';

// build-stage.js
import './build-stage.scss';

// etc.
```
But the current pattern is more or less one big file list in `public/components` so I don't necessarily want to go out of the norm.

Any thoughts? @rhamilto @sg00dwin 
As well as @alecmerdler @spadgett since its a structure question.

Otherwise, I'm going to remove the `[WIP]` because we might be close to settled.

File: frontend/public/components/build-pipeline.scss (hunk 17 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: tab to space fixed.

File: frontend/public/components/build-pipeline.scss (hunk 18 above)
Comment (sg00dwin) [reply frontend/public/components/build-pipeline.scss]: @benjaminapetersen Well, we can split it up to make it smaller, though a cursory look at the code, if you pull the rest out into a `build-pipeline.scss` partial,  the majority of code would still be in `build-stage.scss`. Probably would still be 200 lines.

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.scss]: Yeah I think our verbals on this indicated we don't want to split it up at this point.

File: frontend/public/components/build-pipeline.tsx [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+
+const getBuildNumber = (resource: K8sResourceKind): number => _.get(resource, ['metadata', 'annotations', 'openshift.io/build.number']);
+const getStages = (status): any[] => (status && status.stages) || [];
+const getJenkinsStatus = (resource: K8sResourceKind) => {
+  const status = _.get(resource, ['metadata', 'annotations', 'openshift.io/jenkins-status-json']);
+  if (!status) {
+    return {};
+  }
+
+  const json = _.attempt(JSON.parse, status);
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: nit: It's no longer JSON now that you've parsed it... I'd swap these variable names.

```js
const json = _.get(resource, ['metadata', 'annotations', 'openshift.io/jenkins-status-json']);
// ...
const status = _.attempt(JSON.parse, json);
```

File: frontend/public/components/build-pipeline.tsx [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+    running: 'fa-refresh fa-spin',
+    complete: 'fa-check-circle',
+    failed: 'fa-times-circle'
+  })[statusClass];
+  const hide = icon ? 'hide' : '';
+  return <React.Fragment>
+    <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
+      <span aria-hidden="true">
+        <span className={`fa ${icon} fa-fw`}></span>
+      </span>
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: Why the extra span? Couldn't this be...

```jsx
<span className={`fa ${icon} fa-fw`} aria-hidden="true"></span>
```

File: frontend/public/components/build-pipeline.tsx [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+const BuildSummaryStatusIcon: React.SFC<BuildSummaryStatusIconProps> = ({status}) => {
+  const statusClass = _.lowerCase(status);
+  const icon = ({
+    new: 'fa-hourglass-o',
+    pending: 'fa-hourglass-half',
+    running: 'fa-refresh fa-spin',
+    complete: 'fa-check-circle',
+    failed: 'fa-times-circle'
+  })[statusClass];
+  const hide = icon ? 'hide' : '';
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: Why use the `hide` class if we could just use JSX to hide the node. Is there some animation we want?

File: frontend/public/components/build-pipeline.tsx [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+  return <React.Fragment>
+    <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
+      <span aria-hidden="true">
+        <span className={`fa ${icon} fa-fw`}></span>
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: We check for empty icon above. Doesn't that mean this could result in `class="fa fa-fw"` with no icon here?

File: frontend/public/components/build-pipeline.tsx [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+
+const BuildLogLink: React.SFC<BuildLogLinkProps> = ({obj}) => {
+  const link = getJenkinsLogURL(obj);
+  return link ? <div className="build-pipeline__link">
+    <a href={link} className="build-pipeline__log-link" target="_blank" rel="noopener noreferrer">View Log</a>
+  </div> : null;
+};
+
+const StagesNotStarted: React.SFC<StagesNotStartedProps> = ({stages}) => {
+  return !stages.length ? <div className="build-pipeline__stage build-pipeline__stage--none">No stages have started.
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: Prefer _.isEmpty

File: frontend/public/components/build-pipeline.tsx [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+    <div className="build-pipeline__circle-clip2"></div>
+    <div className="build-pipeline__circle-inner">
+      <div className="build-pipeline__circle-inner-fill"></div>
+    </div>
+  </div>
+</div>;
+
+const JenkinsInputUrl: React.SFC<JenkinsInputUrlProps> = ({obj, stage}) => {
+  const pending = stage && (stage.status === 'PAUSED_PENDING_INPUT');
+  const buildUrl = getJenkinsBuildURL(obj);
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: nit: Return early if not pending.

```js
const pending = stage && (stage.status === 'PAUSED_PENDING_INPUT');
if (!pending) {
  return;
}

const buildUrl = getJenkinsBuildURL(obj);
return <div className="build-pipeline__stage-actions text-muted">
  <a href={buildUrl} target="_blank" rel="noopener noreferrer">Input Required</a>
</div>;
```

File: frontend/public/components/build-pipeline.tsx [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+  }
+  return null;
+};
+
+const BuildStageTimestamp:React.SFC<BuildStageTimestampProps> = ({timestamp}) => <div className="build-pipeline__stage-time text-muted">
+  { fromNow(timestamp) }
+</div>;
+
+const BuildStageName: React.SFC<BuildStageNameProps> = ({name}) => {
+  return <div title={name} className="build-pipeline__stage-name">
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: `title` is because long names get elided?

File: frontend/public/components/build-pipeline.tsx [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+    <div className="build-pipeline__circle-clip1"></div>
+    <div className="build-pipeline__circle-clip2"></div>
+    <div className="build-pipeline__circle-inner">
+      <div className="build-pipeline__circle-inner-fill"></div>
+    </div>
+  </div>
+</div>;
+
+const JenkinsInputUrl: React.SFC<JenkinsInputUrlProps> = ({obj, stage}) => {
+  const pending = stage && (stage.status === 'PAUSED_PENDING_INPUT');
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: I don't think you need to check if stage is defined here. We don't check other places, so it will be an error before it gets to this point.

File: frontend/public/components/build-pipeline.tsx [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+export const BuildPipeline: React.SFC<BuildPipelineProps> = ({obj}) => {
+  const jenkinsStatus: any = getJenkinsStatus(obj);
+  const stages = getStages(jenkinsStatus);
+  return <div className="build-pipeline">
+    <BuildPipelineSummary obj={obj} />
+    <div className="build-pipeline__container">
+      <div className="build-pipeline__stages">
+        <StagesNotStarted stages={stages} />
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: This is a little confusing because it looks like `StagesNotStarted` is always visible, but then you loop over stages below. I think it's much clearer if you didn't use the component:

```jsx
{_.isEmpty(stages)
  ? <div className="build-pipeline__stage build-pipeline__stage--none">No stages have started.</div>
  : stages.map(stage => <BuildStage obj={obj} stage={stage} key={stage.id} /> )}
```

File: frontend/public/components/build.tsx (hunk 4 above)
Comment (spadgett) [reply frontend/public/components/build.tsx]: Maybe just check `build.spec.strategy.type === 'JenkinsPipeline'`?

File: frontend/public/components/build-pipeline.tsx [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+        { stages.map((stage => <BuildStage obj={obj} stage={stage} key={stage.id} /> )) }
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: Why the double parens (( )) ?

**Comment by benjaminapetersen** [comment]
Addressing these & pushing an update shortly.

File: frontend/public/components/build.tsx (hunk 4 above)
Comment (benjaminapetersen) [reply frontend/public/components/build.tsx]: Yeah, def plenty reasonable. Lets go that route.

File: frontend/public/components/build-pipeline.tsx (hunk 21 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: @sg00dwin there isn't a `.hide` class in the `build-pipelines.scss` file, doing a `grep -ri 'hide' public/components'` doesn't reveal a generic hide class with an animation, so I'll update this to use `jsx`.

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Will be:

```html
{!icon && <span className="build-pipeline__status-icon">
      <span className="fa fa-refresh fa-spin" aria-hidden="true"></span>
</span>}
```

File: frontend/public/components/build-pipeline.tsx (hunk 22 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Probably should be:

```html
  return <React.Fragment>
    {icon && <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
      <span className={`fa ${icon} fa-fw`} aria-hidden="true"></span>
    </span>}

    {!icon && <span className="build-pipeline__status-icon">
      <span className="fa fa-refresh fa-spin" aria-hidden="true"></span>
    </span>}
  </React.Fragment>;
```

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Then the `<React.Fragment>` is not really needed, we will either do one or the other...

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: So probably should be:

```html
  return icon
    ? <span className={`build-pipeline__status-icon build-pipeline__status-icon--${statusClass}`}>
      <span className={`fa ${icon} fa-fw`} aria-hidden="true"></span>
    </span>
    : <span className="build-pipeline__status-icon">
      <span className="fa fa-refresh fa-spin" aria-hidden="true"></span>
    </span>
```

File: frontend/public/components/build-pipeline.tsx (hunk 21 above)
Comment (spadgett) [reply frontend/public/components/build-pipeline.tsx]: Yeah, suggest an sr-only span in there too for screen readers

File: frontend/public/components/build-pipeline.tsx (hunk 24 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Will do.

File: frontend/public/components/build-pipeline.tsx (hunk 20 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Sure can, updating.

File: frontend/public/components/build-pipeline.tsx (hunk 19 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Got it.

File: frontend/public/components/build-pipeline.tsx (hunk 25 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Think so:

```css
.build-pipeline__stage-name {
  @include text-overflow();
  margin-bottom: $pipeline-padding + 3px;
}
```

File: frontend/public/components/build-pipeline.tsx (hunk 26 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: Makes sense. hopefully the check in `build.tsx` means we won't start rendering any of this unless there are jenkins pipeline stages.

File: frontend/public/components/build-pipeline.tsx (hunk 27 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: I can put the `_.isEmpty()` check up here, that makes sense.

File: frontend/public/components/build-pipeline.tsx (hunk 28 above)
Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: I was just envious of Lisp the day I wrote this...

File: frontend/public/components/build-pipeline.tsx (hunk 25 above)
Comment (sg00dwin) [reply frontend/public/components/build-pipeline.tsx]: that's correct

File: frontend/public/components/build-pipeline.tsx [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+      <span className={`fa ${icon} fa-fw`} aria-hidden="true"></span>
+    </span>
+    : <span className="build-pipeline__status-icon">
+      <span className="fa fa-refresh fa-spin" aria-hidden="true"></span>
+    </span>;
+};
+
+const BuildLogLink: React.SFC<BuildLogLinkProps> = ({ obj }) => {
+  const link = getJenkinsLogURL(obj);
+  return link ? <div className="build-pipeline__link">
```
Comment (spadgett) [inline frontend/public/components/build-pipeline.tsx]: nit: does this work?

```jsx
return link && <div> ... </div>;
```

Comment (benjaminapetersen) [reply frontend/public/components/build-pipeline.tsx]: I bet it would have!  Woops.
